metrics.go

```diff
@@ -0,0 +1,268 @@
+//
+// MinIO Object Storage (c) 2022 MinIO, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package madmin
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// MetricType is a bitfield representation of different metric types.
+type MetricType uint32
+
+const (
+	MetricsNone    MetricType = 0
+	MetricsScanner MetricType = 1 << (iota)
+
+	// MetricsAll must be last.
+	// Enables all metrics.
+	MetricsAll = 1<<(iota) - 1
+)
+
+// MetricsOptions are options provided to Metrics call.
+type MetricsOptions struct {
+	Type     MetricType    // Return only these metric types. Several types can be combined using |. Leave at 0 to return all.
+	N        int           // Maximum number of samples to return. 0 will return endless stream.
+	Interval time.Duration // Interval between samples. Will be rounded up to 1s.
+	Hosts    []string      // Leave empty for all
+	ByHost   bool          // Return metrics by host.
+}
+
+// Metrics makes an admin call to retrieve metrics.
+// The provided function is called for each received entry.
+func (adm *AdminClient) Metrics(ctx context.Context, o MetricsOptions, out func(RealtimeMetrics)) (err error) {
+	path := fmt.Sprintf(adminAPIPrefix + "/metrics")
+	q := make(url.Values)
+	q.Set("types", strconv.FormatUint(uint64(o.Type), 10))
+	q.Set("n", strconv.Itoa(o.N))
+	q.Set("interval", o.Interval.String())
+	q.Set("hosts", strings.Join(o.Hosts, ","))
+	if o.ByHost {
+		q.Set("by-host", "true")
+	}
+	resp, err := adm.executeMethod(ctx,
+		http.MethodGet, requestData{
+			relPath:     path,
+			queryValues: q,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		closeResponse(resp)
+		return httpRespToErrorResponse(resp)
+	}
+	defer closeResponse(resp)
+	dec := json.NewDecoder(resp.Body)
+	for {
+		var m RealtimeMetrics
+		err := dec.Decode(&m)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				err = io.ErrUnexpectedEOF
+			}
+			return err
+		}
+		out(m)
+		if m.Final {
+			break
+		}
+	}
+	return nil
+}
+
+// Contains returns whether m contains all of x.
+func (m MetricType) Contains(x MetricType) bool {
+	return m&x == x
+}
+
+// RealtimeMetrics provides realtime metrics.
+// This is intended to be expanded over time to cover more types.
+type RealtimeMetrics struct {
```

Comment:
maybe AggregatedMetrics seems to be a better name here

Reply:
I think `RealtimeMetrics` is more specific. I will take inspiration and change the field "Total" to "Aggregated".